crates/stages/src/stages/merkle.rs

```diff
@@ -70,15 +120,407 @@ impl<DB: Database> Stage<DB> for MerkleStage {
+            stage_progress: Some(stage_progress),
+        };
+
+        runner.seed_execution(input).expect("failed to seed execution");
+
+        let rx = runner.execute(input);
+
+        // Assert the successful result
+        let result = rx.await.unwrap();
+        assert_matches!(
+            result,
+            Ok(ExecOutput { done, stage_progress })
+                if done && stage_progress == previous_stage
+        );
+
+        // Validate the stage execution
+        assert!(runner.validate_execution(input, result.ok()).is_ok(), "execution validation");
+    }
+
+    struct MerkleTestRunner {
+        tx: TestTransaction,
+        clean_threshold: u64,
+    }
+
+    impl Default for MerkleTestRunner {
+        fn default() -> Self {
+            Self { tx: TestTransaction::default(), clean_threshold: 100000 }
+        }
+    }
+
+    impl StageTestRunner for MerkleTestRunner {
+        type S = MerkleStage;
+
+        fn tx(&self) -> &TestTransaction {
+            &self.tx
+        }
+
+        fn stage(&self) -> Self::S {
+            Self::S::Both { clean_threshold: self.clean_threshold }
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl ExecuteStageTestRunner for MerkleTestRunner {
+        type Seed = Vec<SealedBlock>;
+
+        fn seed_execution(&mut self, input: ExecInput) -> Result<Self::Seed, TestRunnerError> {
+            let stage_progress = input.stage_progress.unwrap_or_default();
+            let end = input.previous_stage_progress() + 1;
+
+            let n_accounts = 31;
+            let mut accounts = random_contract_account_range(&mut (0..n_accounts));
+
+            let SealedBlock { header, body, ommers } =
+                random_block(stage_progress, None, Some(0), None);
+            let mut header = header.unseal();
+            header.state_root = self.generate_initial_trie(&accounts)?;
+            let sealed_head = SealedBlock { header: header.seal(), body, ommers };
+
+            let head_hash = sealed_head.hash();
+            let mut blocks = vec![sealed_head];
+
+            blocks.extend(random_block_range((stage_progress + 1)..end, head_hash, 0..3));
+
+            self.tx.insert_headers(blocks.iter().map(|block| &block.header))?;
+
+            let (mut transition_id, mut tx_id) = (0, 0);
+
+            let mut storages: BTreeMap<Address, BTreeMap<H256, U256>> = BTreeMap::new();
+
+            for progress in blocks.iter() {
+                // Insert last progress data
+                self.tx.commit(|tx| {
+                    let key: BlockNumHash = (progress.number, progress.hash()).into();
+
+                    let body = StoredBlockBody {
+                        start_tx_id: tx_id,
+                        tx_count: progress.body.len() as u64,
+                    };
+
+                    progress.body.iter().try_for_each(|transaction| {
+                        tx_id += 1;
+                        transition_id += 1;
```

Comment:
You would need to move these increments to the bottom of the loop. and index would start from zero that way.